Support external compilers

gnidan opened the pull request:
Ref: #1134 

# Support external compilers

## Motivation

It'd be nice to use Truffle for workflows that don't operate nicely in the context of the current `truffle compile` solc pipeline. This compilation step can be greatly generalized by way of offering configuration for two components:
- Allow specifying arbitrary command to run, for compilation
- Allow specifying targets, with optional arbitrary command to process those targets into Truffle artifacts.

This PR makes changes to `truffle-workflow-compile` to support multiple compilers (at all), and defines a new package `truffle-external-compile`, which implements the above.

## Approach

In `truffle-workflow-compile`:

- Define `SUPPORTED_COMPILERS`, mapping name to node module

- Detect keys included in `config.compilers` (NOTE: solc is defaulted!)

- Run all compile modules for detected supported compilers

- Add `--compiler` option to the `truffle compile` command, to specify a particular compiler instead.

- Update truffle-workflow-compile's interface to return an object of the form:
  ```javascript
  {
    contracts, // aggregate of all contracts by name
    outputs: {
      compilerName: compilerSpecificResult
      /* for instance: */
      solc: paths
    }
  }
  ``` 
  This is instead of having multiple return values via `callback(err, contracts, paths)`.

New package `truffle-external-compile`:

- See [Configuration Example](https://github.com/trufflesuite/truffle/blob/d6b957ba9d7a7468de3559bf5e387e9b2e483b06/packages/truffle/test/sources/external_compile/truffle.js#L2-L22)

- Define method for processing compilation targets via arbitrary commands:

  1. Run `compilers.external.command`, expect some target output

  2. For each defined target, glob-expand `target.path` and pipe to `target.command` via stdin.

  3. Use stdout from each `target.command` as JSON description of artifact

- Define method for processing compilation targets via explicit property listing.

### re: globbing in first commit
The first commit here changes the way `truffle-contract-sources` works, so that compiler configuration can do things like:
```javascript
compilers: {
  solc: {
    sources: "**/*.sol",
    /* ... */
  }
}
```

## Dependent on `truffle-workflow-compile`'s external interface

These have all been updated:

- `truffle-core/lib/test` - resolves `contracts` and `paths` in `compileContractsWithTestFilesIfNeeded`
- `truffle-debugger/test/helpers` - resolves `contracts` and `paths` in `compile`
- `truffle-core/test/compile` - resolves `contracts` in three tests
- `truffle-core/test/migrate` - resolves `contracts` in one test
- `truffle-core/test/ethpm` - resolves `contracts` in one test
- `truffle-core/test/npm` - resolves `contracts` in one test

## Known Limitations

- Solidity tests currently don't work properly when using `truffle compile` as external

gnidan commented:
@cgewecke Let me know if I should separate the first "globbing" commit into its own PR. It's not really used right now, so it's not really part of this scope.

cgewecke commented on packages/truffle-external-compile/index.js:
```diff
@@ -0,0 +1,80 @@
+
+const { exec, execSync } = require('child_process');
+const path = require("path");
+const { callbackify, promisify } = require("util");
```
Didn't know callbackify existed!

cgewecke commented on packages/truffle-external-compile/index.js:
```diff
@@ -0,0 +1,80 @@
+
+compile.all = compile;
+compile.necessary = compile;
+
```
👍

cgewecke commented on packages/truffle-workflow-compile/index.js:
```diff
@@ -19,69 +64,68 @@ var Contracts = {
-        callback(err);
-        return;
-      }
+    const result = await promisify(mkdirp)(options.contracts_build_directory);
```
Nice.

cgewecke commented on packages/truffle/test/scenarios/external_compilers/truffle-compile.js:
```diff
@@ -0,0 +1,99 @@
+        return done(err);
+      }
+
+      assert(output.indexOf("3 passing") >= 0);
```
👍

cgewecke commented on packages/truffle/test/sources/external_compile/truffle.js:
```diff
@@ -0,0 +1,22 @@
+        "--contracts_build_directory=external",
+      targets: [{
+        path: "external/*.json",
+        command: "cat"
```
What does `command` do? Is it a post-processing step?

gnidan replied:
Yep - takes whatever files specified by `path` and pipes them to the `command`. The command is expected to output a JSON artifact on stdout.

cgewecke replied:
Sweet.

cgewecke commented on packages/truffle/scripts/test.sh:
```diff
@@ -3,7 +3,7 @@
+  npm run build-cli && mocha --timeout 50000 --grep @ganache --invert --colors $@
 else
-  npm run build-cli && mocha --no-warnings --timeout 7000 --grep @geth --invert --colors
+  npm run build-cli && mocha --no-warnings --timeout 7000 --grep @geth --invert --colors $@
```
What's being added here?

gnidan replied:
It forwards arguments; it's just for doing `yarn test -b` so that the `-b` gets forwarded to mocha.

cgewecke commented on packages/truffle/test/scenarios/external_compilers/truffle-compile.js:
```diff
@@ -0,0 +1,99 @@
+  it("will compile", function(done) {
+    this.timeout(20000);
+
+    CommandRunner.run("compile --compiler=external", config, function(err) {
```
I like this test. Without this flag compile will run both solc compilation *and* whatever is specified by external?

gnidan replied:
Correct

cgewecke commented on packages/truffle/test/scenarios/external_compilers/truffle-compile.js:
```diff
@@ -0,0 +1,99 @@
+      }
+
+      assert(fs.existsSync(path.join(config.contracts_build_directory, "MetaCoin.json")));
+      assert(fs.existsSync(path.join(config.contracts_build_directory, "ConvertLib.json")));
```
The write here is to a folder named `external`? Can we check the path for that string?

gnidan replied:
`truffle compile --compiler=external` calls `truffle compile --compiler=solc --contracts_build_directory=external`, which puts the contracts in `external/`, but the project is configured to still use `build/contracts/` for artifacts. So the external compiler takes what's in `external/`, `cat`s them, and saves them in `build/contracts/`

cgewecke replied:
Ahhh ok thanks makes sense.

gnidan commented:
@cgewecke agreed about all this. I'll add better testing when I get back; I felt rushed trying to add all these options in. It's messy code as a result :) Holding off merging, awaiting required extra love.

(Just published `truffle@external-compiler` for anyone who wants to try this out)

axic commented:
I have tried this now with my Rust environment:
```
module.exports = {
  // See <http://truffleframework.com/docs/advanced/configuration>
  // to customize your Truffle configuration!
  compilers: {
    external: {
      command: "cargo build --release --target wasm32-unknown-unknown",
      targets: [{
        properties: {
          contractName: "ecadd"
        },
        fileProperties: {
          bytecode: "target/wasm32-unknown-unknown/release/ewasm_precompile_ecadd.wasm"
        }
      }]
    }
  }
};
```

Good news is that the command gets invoked and I do get an artifact file.

One note: initially we've discussed that if the `abi` field is omitted  then it will consider it as a ABI with a single fallback function. That is not the case right now.

And also there don't seem to be a validation that the resulting artifact is correct, in my case it only contains `contractName`, `bytecode`, `networks`, `schemaVersion` and `updatedAt` fields.

There seems to be one issue though: the bytecode kind of looks like weirdly escaped.
```
  "bytecode": "0x\u0000asm\u0001\u0000\u0000\u0000\u0001�\u0003@`\u0003^?^?^?\u0000`\u0002^?^?\u0001^?`\u0000\u0000`\u0003^?^?^?\u0001^?`\u0000\u0001^?`\u0001~\u0000`\u0004^?^?^?^?\u0000`\u0001^?\u0000`\u0001^?\u0001^?`\u0004^?^?^?^?\u0001^?`\u0002^?^?\u0000`\u0004^?^?^?~\u0000`\u0005^?^?^?^?^?\u0001^?``
```

It has unicode escapes and unprintable characters. The input is raw binary.

axic commented:
Without having set an alternative RPC, trying to deploy it via ganache of course fails, but there's no useful error message:
```
$ truffle test
Using network 'test'.

    Finished release [optimized] target(s) in 0.09s
Error
    at /Users/alex/.nvm/versions/node/v10.5.0/lib/node_modules/truffle/build/webpack:/packages/truffle-deployer/src/deployment.js:354:1
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

Another tiny issue is that the output from the compile command seems to be weirdly cut up at places. It may be down to the fact that it is coloured (the colours are also removed via truffle):
```
$ truffle compile
   Compiling semver-parser v0.7.0
   Compiling serde v1.0.70
   Compiling heapsize v0.4.2
   Compiling crunchy v0.1.6
   Compiling num-traits v0.2.5
   Compiling num-integer v
0.1.39
   Compiling crunchy v0.2.1
   Compiling rand v0.4.2
   Compiling 
rustc-hex v1.0.0
   Compiling byteorder v1.2.3
   Compiling num-bigint v
0.2.0
   Compiling parity-bytes v0.1.0 (https://github.com/paritytech/parity-common#4481687a)
   Compiling semver v0.9.0
   Compiling tiny-keccak v1.4.2
   Compiling rustc_version
 v0.2.2
   Compiling fixed-hash
 v0.2.1
   Compiling bn v0.4.4 (https://github.com/paritytech/bn#786c0d56)
   Compiling uint v
0.2.1
   Compiling ethereum-types v
0.3.2
   Compiling ewasm_api v
0.1.0 (https://github.com/ewasm/ewasm-rust-api#17b4da5c)
   Compiling ethereum-types-serialize v
0.2.1
   Compiling ethbloom v0.5.0
   Compiling ethereum-bn128
 v0.0.1 (https://github.com/ewasm/ethereum-bn128.rs#4eaba2ce)
   Compiling ewasm-precompile-ecadd v
0.1.0 (file:///Users/alex/Projects/ewasm.rs/ewasm-precompiles/ecadd)
    Finished 
release [optimized] target(s) in 1m 5.26s
Writing artifacts to ./build/contracts
```

axic commented:
Here is my working copy: https://github.com/ewasm/ewasm-precompiles/tree/ecadd-truffle-tests/ecadd

gnidan commented:
Ah thanks for trying it out @axic.

When I get back, I'll:
- fix ABI to have that fallback-only default
- fix bytecode interpretation (I didn't realize the output was actual binary, plain text, silly me!)
- sort out those weird extra newlines

Hopefully you can work around these in the meantime and get a bit further along.

jklepatch commented on packages/truffle-external-compile/index.js:
```diff
@@ -57,15 +57,23 @@ const compile = callbackify(async function(options) {
-        : input;
+    expect.one(target, [ "path", "command" ]);  // also allows both
+
+    if (target.path != undefined) {
```
this is safer to test for undefined: `typeof target.path !== "undefined")`

gnidan replied:
Yep, except this is on purpose to allow `null` as a valid non-value :)

lrettig commented on packages/truffle-external-compile/index.js:
```diff
@@ -0,0 +1,154 @@
+    const contract = Object.assign({}, target.properties || {});
+
+    for (let [key, path] of Object.entries(target.fileProperties || {})) {
+      const contents = fs.readFileSync(resolve(cwd, path)).toString();
```
If you change `toString()` to `toString('hex')` here, it (at least partially) fixes the bytecode issue.

lrettig replied:
Here is a slightly more sophisticated approach: https://github.com/lrettig/truffle/commit/acfc8b35e46b89554df1689d936a7829d068c5e5

axic replied:
I don't think that is a good idea. Why make a distinction for wasm?

lrettig commented:
Another thought: if we want to be thorough about adding support for Ewasm, then instead of using evm properties like `bytecode`, `sourceMap`, etc. it probably makes more sense to use the Wasm properties laid out here: https://github.com/ethereum/solidity/commit/4b5639bf63f2c86604cb55387324e00759a0b1f2#diff-d7b1fb9ae448539124dcfd3d09ad6e60R124 e.g. `wasm`, `wast`, etc. and maybe add these to the Truffle contract schema under `wasm`.

gnidan commented:
@lrettig that's a good point, but out of scope for this effort.

Those properties are the properties that truffle-contract-schema uses, which can be considered coincidentally the same as the Solidity ones. This rename should probably be considered when we get to implementing an overhaul of Truffle's artifacts format.

Would you mind opening an issue about that here? https://github.com/trufflesuite/artifact-updates

Thanks!

axic commented:
@lrettig @gnidan note that isn't fully specified in Solidity, because ewasm isn't fully specified in that regard (whether it would have a deployer code with a constructor or not).

gnidan commented on packages/truffle-external-compile/index.js:
```diff
@@ -39,6 +41,21 @@ const runCommand = promisify(function (command, options, callback) {
+function decodeContents (contents) {
+  // HACK to test if contents has binary data
+  // from https://stackoverflow.com/a/49773659
+  if (/\ufffd/.test(contents.toString())) {
```
Hm, so I was thinking that `contents.toString()` was "correctly" rendering those unreadable characters. Is that not the case in all situations?

Do you have any other ideas as to how to detect this, @axic? I'm trying to avoid making these files parsed in a property-specific way, and also trying to avoid requiring the user to specify the encoding on a per-property basis.

axic replied:
I would suggest that if the file starts with `0x` (and passes the hex test for the rest, e.g. has only hex characters) then read it as a hex string otherwise consider it a binary.

gnidan replied:
Mmm. OK so how about:

- If it parses as JSON, it's JSON
- If it starts with `0x`, it's hex
- Otherwise, it's binary

axic commented:
To summarise my initial reported issues:
- [x] optional abi field (falls back to a default one)
- [x] binary bytecode is handled wrongly
- [ ] compile command output is mangled
- [ ] validate that the resulting artifact file has all the necessary fields
- [ ] no real error message if deployment fails

Should I open issues for some of these or is it planned to be handled in this PR?

gnidan commented:
@axic I'm going to include those fixes in this PR.

Should have those resolved tomorrow and I'll put out a new build for you to try. Thanks!

axic commented:
Great, thank you!

gnidan commented:
@axic I just published `truffle@5.0.0-external-compiler.1`. I think I addressed all your concerns. Could you please try it out and let me know? Thanks!

N.B. when you install this version, you will likely need to uninstall first:
```
npm uninstall -g truffle
npm install -g truffle@5.0.0-external-compiler.1
```

axic commented:
Test works nicer:
```
$ truffle test
Could not connect to your Ethereum client. Please check that your Ethereum client:
    - is running
    - is accepting RPC connections (i.e., "--rpc" option is used in geth)
    - is accessible over the network
    - is properly configured in your Truffle configuration file (truffle.js)
```

The line mangling still appears:
```
$ truffle compile
   Compiling semver-parser v0.
7.0
   Compiling serde v1.0.70
   Compiling crunchy v0.1.6
   Compiling heapsize v0.4.2
   Compiling num-traits v0.2.5
   Compiling rustc-hex v1
.0.0
   Compiling rand v
0.4.2
   Compiling num-integer
 v0.1.39
```

The default ABI is populated and the binary seems to be picked up properly. Nice!

gnidan commented:
@cgewecke any idea why the `deployFailed` event error isn't showing up nicely?

cgewecke commented:
@gnidan Ai it's a bug. I thought I fixed it in #1113 :/ 

Tracking down cause rn but it affects all tests so it shouldn't block this PR. Will open something shortly.

cgewecke commented on packages/truffle-external-compile/index.js:
```diff
@@ -0,0 +1,279 @@
+    }
+  }
+}
+
```
Nice

cgewecke commented on packages/truffle-external-compile/test/test_targets.js:
```diff
@@ -0,0 +1,122 @@
+    });
+
+    it("reads UTF-8 file properties", async () => {
+      const contractName = "My😀Contract";
```
lol!

cgewecke commented:
@gnidan Looks good!

axic commented:
> Output is now properly buffered and should appear without extraneous linebreaks :)

Confirmed, it looks nice!

axic commented:
Also regarding validation this is what I've got when I've removed the bytecode field from the config:
```
 Warning: contract ecadd does not specify bytecode. You won't be able to deploy it.
```

I guess validation works 👍

gnidan commented:
@axic that's some explicit validation, because technically `bytecode` isn't required as part of truffle-contract-schema. But the actual schema validation should work, too. (Try using a malformed ABI ;)

Anything else before I merge and call this a wrap for now?

Thanks for all the good feedback!

axic commented:
I haven't gotten around to test it with a native ewasm client with RPC (it will be cpp-ethereum/aleth) - I may have some more things to report then.

However that may not happen before Tuesday, so I'd say go ahead.

gnidan commented:
Great! I'll merge and get you a new `truffle@next` release to use for your testing.
